File: app/src/main/java/com/medo/tweetspie/main/MainActivity.java
package com.medo.tweetspie.main;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ProgressBar;

import com.medo.tweetspie.R;
import com.medo.tweetspie.base.BaseActivity;
import com.medo.tweetspie.bus.events.TimelineServiceEvent;
import com.medo.tweetspie.database.RealmModule;
import com.medo.tweetspie.database.model.RealmTweet;
import com.medo.tweetspie.injection.components.AppComponent;
import com.medo.tweetspie.injection.components.DaggerUserComponent;
import com.medo.tweetspie.main.adapter.AdapterContract;
import com.medo.tweetspie.main.adapter.AdapterPresenter;
import com.medo.tweetspie.main.adapter.TweetsAdapter;
import com.medo.tweetspie.main.viewer.ViewerActivity;
import com.medo.tweetspie.onboarding.OnboardingActivity;
import com.medo.tweetspie.rest.TwitterModule;
import com.medo.tweetspie.service.TimelineService;
import com.medo.tweetspie.utils.DividerItemDecoration;
import com.medo.tweetspie.utils.IntentUtils;
import com.squareup.otto.Subscribe;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.realm.OrderedRealmCollection;


public class MainActivity extends BaseActivity implements MainContract.View, AdapterContract.View {

  private static final int REQUEST_CODE_ONBOARDING = 1620;

  @BindView(R.id.recycler_tweets)
  RecyclerView recyclerTweets;
  @BindView(R.id.progress)
  ProgressBar progress;

  @Inject
  MainPresenter presenter;
  @Inject
  AdapterPresenter adapterPresenter;

  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {

    super.onCreate(savedInstanceState);
    presenter.onAttach(this);
  }

  @Override
  protected void onDestroy() {

    super.onDestroy();
    presenter.onDetach();
    adapterPresenter.onDetach();
  }

  @Override
  protected void onActivityResult(int requestCode, int resultCode, Intent data) {

    super.onActivityResult(requestCode, resultCode, data);
    if (requestCode == REQUEST_CODE_ONBOARDING) {
      if (resultCode == RESULT_OK) {
        presenter.onOnboardingSuccess();
      }
      else {
        presenter.onOnboardingFailure();
      }
    }
  }

  @Override
  public void startOnboarding() {

    startActivityForResult(OnboardingActivity.getIntent(this), REQUEST_CODE_ONBOARDING);
  }

  @Override
  protected void inject(@NonNull AppComponent appComponent) {

    DaggerUserComponent.builder()
            .appComponent(appComponent)
            .mainModule(new MainModule())
            .realmModule(new RealmModule())
            .twitterModule(new TwitterModule())
            .build()
            .inject(this);
  }

  @Override
  public void initUi() {

    setContentView(R.layout.activity_main);
    ButterKnife.bind(this);
  }

  @Override
  public void loadData() {

    // show progress bar and start loading data
    progress.setVisibility(View.VISIBLE);
    TimelineService.start(this);
  }

  @Override
  public void showData(OrderedRealmCollection<RealmTweet> tweets) {

    // hide the progress bar
    progress.setVisibility(View.GONE);

    // create an adapter for displaying tweets
    TweetsAdapter tweetsAdapter = (TweetsAdapter) recyclerTweets.getAdapter();
    if (tweetsAdapter != null) {
      // just update the data if the adapter has already been set
      tweetsAdapter.updateData(tweets);
      return;
    }

    adapterPresenter.onAttach(this);
    // setup the recycler view and set the adapter
    tweetsAdapter = new TweetsAdapter(this, tweets, adapterPresenter);
    recyclerTweets.setLayoutManager(new LinearLayoutManager(this));
    recyclerTweets.setAdapter(tweetsAdapter);
    recyclerTweets.setHasFixedSize(true);
    recyclerTweets.addItemDecoration(new DividerItemDecoration(this, DividerItemDecoration.VERTICAL_LIST));
  }

  @Override
  public void showError() {

    // TODO show error
  }

  @Override
  public void exit() {

    finish();
  }

  @Override
  public void openTweetMedia(@NonNull String tweetId) {

    final Intent intent = ViewerActivity.getIntent(this, tweetId);
    startActivity(intent);
  }

  @Override
  public void openUrl(@NonNull String url) {

    IntentUtils.openUrl(this, url);
  }

  @Subscribe
  public void onTimelineServiceResult(TimelineServiceEvent event) {

    presenter.onDataLoaded(event.isSuccess());
  }
}

File: app/src/main/java/com/medo/tweetspie/main/MainActivity.kt
package com.medo.tweetspie.main

import android.app.Activity
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import com.medo.tweetspie.R
import com.medo.tweetspie.base.BaseActivity
import com.medo.tweetspie.di.BindContext
import com.medo.tweetspie.extensions.launchActivity
import com.medo.tweetspie.extensions.register
import com.medo.tweetspie.extensions.toast
import com.medo.tweetspie.extensions.unregister
import com.medo.tweetspie.onboarding.OnboardingActivity
import com.medo.tweetspie.service.MainService
import com.medo.tweetspie.utils.ACTION_UPDATE
import com.medo.tweetspie.utils.DividerItemDecoration
import com.medo.tweetspie.utils.EXTRA_SUCCESS
import com.medo.tweetspie.utils.REQUEST_CODE_ONBOARDING
import kotlinx.android.synthetic.main.activity_main.*
import org.koin.android.ext.android.inject


class MainActivity : BaseActivity(), MainMvp.View {

    override val contextName: String = BindContext.Main
    override val presenter by inject<MainMvp.Presenter>()

    private val updateReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            presenter.onDataLoaded(intent.getBooleanExtra(EXTRA_SUCCESS, false))
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        presenter.onAttach(this)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode != REQUEST_CODE_ONBOARDING) return
        if (resultCode == Activity.RESULT_OK) presenter.onOnboardingSuccess()
        else presenter.onOnboardingFailure()
    }

    override fun onResume() {
        super.onResume()
        register(updateReceiver, IntentFilter(ACTION_UPDATE))
    }

    override fun onPause() {
        super.onPause()
        unregister(updateReceiver)
    }

    override fun startOnboarding() = launchActivity<OnboardingActivity>(REQUEST_CODE_ONBOARDING)

    override fun loadData() {
        progress.visibility = View.VISIBLE
        startService(Intent(this, MainService::class.java))
    }

    override fun showData(data: List<Any>, cached: Boolean) {
        if (!cached) progress.visibility = View.GONE

        recycler.layoutManager = LinearLayoutManager(this)
        recycler.setHasFixedSize(true)
        recycler.addItemDecoration(DividerItemDecoration(this))
    }

    override fun showError(message: String) = toast(message)

    override fun exit() = finish()
}
